{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4abd08cc-8c4b-44ce-8f72-76c55c64a60b",
   "metadata": {},
   "source": [
    "# Notebook for DE Analysis of single-cell data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d87b9d13-3122-4c6a-8efb-a74d7b619569",
   "metadata": {},
   "source": [
    "## Load Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ebdc2f3-ab7a-4a3e-bd58-7cb54050e10f",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(Seurat))\n",
    "suppressPackageStartupMessages(library(DESeq2))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressPackageStartupMessages(library(stringr))\n",
    "suppressMessages(library(data.table))\n",
    "suppressMessages(library(Matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3185760-c2c4-4059-9396-7ec9dfd82c16",
   "metadata": {},
   "source": [
    "## Define Inputs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8738b67-9bc9-41c7-aec8-0d26221fa7b4",
   "metadata": {},
   "source": [
    "### Assumptions:\n",
    "- The cell type column in your seurat object is called 'cell_type'\n",
    "- The sample/donor ID in your seurat object is called 'library'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ff01e731-8145-45ea-a7ce-e8c9928cb419",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inputs: Seurat object and metadata\n",
    "seurat_obj_path <- '/nfs/lab/sara/projects/PanKbase/singlecell/input/2023_01_27_hpap_res0p3.rds'\n",
    "metadata_path <- '/nfs/lab/sara/projects/PanKbase/singlecell/input/Donor_Summary_127.csv'\n",
    "\n",
    "# Load the Seurat object and metadata\n",
    "data <- readRDS(seurat_obj_path)\n",
    "meta <- read.csv(metadata_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6fdd6fd-c42e-4a14-aa86-62f39fb0e46f",
   "metadata": {},
   "source": [
    "#### Format Inputs - Optional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c06981a9-377f-47a8-8bc7-d6886ff130c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TO USER - Ensure that the metadata has only the donors present in the seurat object - no more no less. Modify or comment out this chunk according to your needs\n",
    "meta <- read.csv(metadata_path)\n",
    "meta <- meta[meta$donor_ID %in% unique(data@meta.data$library), ]\n",
    "\n",
    "dib_stat  <- data@meta.data[,c('library', 'Diabetes_Status_w_AAB', 'chemistry', 'tissue_source')]\n",
    "dib_stat <- unique(dib_stat)\n",
    "rownames(dib_stat) <- NULL\n",
    "\n",
    "meta <- merge(meta, dib_stat, by.x = 'donor_ID', by.y='library')\n",
    "\n",
    "meta <- rename(meta, diabetes_status = Diabetes_Status_w_AAB)\n",
    "meta <- rename(meta, sex = gender)\n",
    "\n",
    "rownames(meta) <- meta$donor_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2f80c28e-d26b-45ed-bad7-c94e4d0131f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 86</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>donor_ID</th><th scope=col>rrid</th><th scope=col>recovery_opo</th><th scope=col>donor_ID_second</th><th scope=col>unos_ID</th><th scope=col>sex</th><th scope=col>bmi</th><th scope=col>age_years</th><th scope=col>race</th><th scope=col>hba1c</th><th scope=col>⋯</th><th scope=col>beta_cell_count</th><th scope=col>delta_cell_count</th><th scope=col>epsilon_cell_count</th><th scope=col>pp_cell_count</th><th scope=col>total_events</th><th scope=col>weight</th><th scope=col>height</th><th scope=col>diabetes_status</th><th scope=col>chemistry</th><th scope=col>tissue_source</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>HPAP-019</th><td>HPAP-019</td><td>RRID:SAMN19776450</td><td></td><td>17D0406</td><td>NA</td><td>Male  </td><td>29.80</td><td>22</td><td>Caucasian</td><td>5.2</td><td>⋯</td><td>38637</td><td> 3327</td><td>180</td><td> 2114</td><td> 55007</td><td>NA</td><td>NA</td><td>AAB+</td><td>10Xv2</td><td>nPod </td></tr>\n",
       "\t<tr><th scope=row>HPAP-020</th><td>HPAP-020</td><td>RRID:SAMN19776451</td><td></td><td>18J0307</td><td>NA</td><td>Male  </td><td>13.20</td><td>14</td><td>Caucasian</td><td> NA</td><td>⋯</td><td>  249</td><td>14485</td><td>939</td><td> 6706</td><td>202889</td><td>NA</td><td>NA</td><td>T1D </td><td>10Xv2</td><td>nPod </td></tr>\n",
       "\t<tr><th scope=row>HPAP-021</th><td>HPAP-021</td><td>RRID:SAMN19776452</td><td></td><td>18M1122</td><td>NA</td><td>Female</td><td>21.40</td><td>13</td><td>Caucasian</td><td> NA</td><td>⋯</td><td>   76</td><td> 3108</td><td>306</td><td> 1897</td><td>  8954</td><td>NA</td><td>NA</td><td>T1D </td><td>10Xv2</td><td>nPod </td></tr>\n",
       "\t<tr><th scope=row>HPAP-022</th><td>HPAP-022</td><td>RRID:SAMN19776453</td><td></td><td>18M2405</td><td>NA</td><td>Female</td><td>34.70</td><td>39</td><td>Caucasian</td><td>4.7</td><td>⋯</td><td> 2788</td><td>34375</td><td>437</td><td> 2070</td><td>315113</td><td>NA</td><td>NA</td><td>ND  </td><td>10Xv2</td><td>UPenn</td></tr>\n",
       "\t<tr><th scope=row>HPAP-023</th><td>HPAP-023</td><td>RRID:SAMN19776454</td><td></td><td>18A0312</td><td>NA</td><td>Female</td><td>21.35</td><td>17</td><td>Caucasian</td><td>8.9</td><td>⋯</td><td>  607</td><td> 3659</td><td>448</td><td>25351</td><td> 31682</td><td>NA</td><td>NA</td><td>T1D </td><td>10Xv2</td><td>nPod </td></tr>\n",
       "\t<tr><th scope=row>HPAP-024</th><td>HPAP-024</td><td>RRID:SAMN19776455</td><td></td><td>18A0813</td><td>NA</td><td>Male  </td><td>24.30</td><td>18</td><td>Caucasian</td><td>5.5</td><td>⋯</td><td>29407</td><td> 2183</td><td>307</td><td> 7435</td><td> 42501</td><td>NA</td><td>NA</td><td>AAB+</td><td>10Xv2</td><td>nPod </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 86\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & donor\\_ID & rrid & recovery\\_opo & donor\\_ID\\_second & unos\\_ID & sex & bmi & age\\_years & race & hba1c & ⋯ & beta\\_cell\\_count & delta\\_cell\\_count & epsilon\\_cell\\_count & pp\\_cell\\_count & total\\_events & weight & height & diabetes\\_status & chemistry & tissue\\_source\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <lgl> & <chr> & <dbl> & <int> & <chr> & <dbl> & ⋯ & <int> & <int> & <int> & <int> & <int> & <dbl> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tHPAP-019 & HPAP-019 & RRID:SAMN19776450 &  & 17D0406 & NA & Male   & 29.80 & 22 & Caucasian & 5.2 & ⋯ & 38637 &  3327 & 180 &  2114 &  55007 & NA & NA & AAB+ & 10Xv2 & nPod \\\\\n",
       "\tHPAP-020 & HPAP-020 & RRID:SAMN19776451 &  & 18J0307 & NA & Male   & 13.20 & 14 & Caucasian &  NA & ⋯ &   249 & 14485 & 939 &  6706 & 202889 & NA & NA & T1D  & 10Xv2 & nPod \\\\\n",
       "\tHPAP-021 & HPAP-021 & RRID:SAMN19776452 &  & 18M1122 & NA & Female & 21.40 & 13 & Caucasian &  NA & ⋯ &    76 &  3108 & 306 &  1897 &   8954 & NA & NA & T1D  & 10Xv2 & nPod \\\\\n",
       "\tHPAP-022 & HPAP-022 & RRID:SAMN19776453 &  & 18M2405 & NA & Female & 34.70 & 39 & Caucasian & 4.7 & ⋯ &  2788 & 34375 & 437 &  2070 & 315113 & NA & NA & ND   & 10Xv2 & UPenn\\\\\n",
       "\tHPAP-023 & HPAP-023 & RRID:SAMN19776454 &  & 18A0312 & NA & Female & 21.35 & 17 & Caucasian & 8.9 & ⋯ &   607 &  3659 & 448 & 25351 &  31682 & NA & NA & T1D  & 10Xv2 & nPod \\\\\n",
       "\tHPAP-024 & HPAP-024 & RRID:SAMN19776455 &  & 18A0813 & NA & Male   & 24.30 & 18 & Caucasian & 5.5 & ⋯ & 29407 &  2183 & 307 &  7435 &  42501 & NA & NA & AAB+ & 10Xv2 & nPod \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 86\n",
       "\n",
       "| <!--/--> | donor_ID &lt;chr&gt; | rrid &lt;chr&gt; | recovery_opo &lt;chr&gt; | donor_ID_second &lt;chr&gt; | unos_ID &lt;lgl&gt; | sex &lt;chr&gt; | bmi &lt;dbl&gt; | age_years &lt;int&gt; | race &lt;chr&gt; | hba1c &lt;dbl&gt; | ⋯ ⋯ | beta_cell_count &lt;int&gt; | delta_cell_count &lt;int&gt; | epsilon_cell_count &lt;int&gt; | pp_cell_count &lt;int&gt; | total_events &lt;int&gt; | weight &lt;dbl&gt; | height &lt;dbl&gt; | diabetes_status &lt;chr&gt; | chemistry &lt;chr&gt; | tissue_source &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HPAP-019 | HPAP-019 | RRID:SAMN19776450 | <!----> | 17D0406 | NA | Male   | 29.80 | 22 | Caucasian | 5.2 | ⋯ | 38637 |  3327 | 180 |  2114 |  55007 | NA | NA | AAB+ | 10Xv2 | nPod  |\n",
       "| HPAP-020 | HPAP-020 | RRID:SAMN19776451 | <!----> | 18J0307 | NA | Male   | 13.20 | 14 | Caucasian |  NA | ⋯ |   249 | 14485 | 939 |  6706 | 202889 | NA | NA | T1D  | 10Xv2 | nPod  |\n",
       "| HPAP-021 | HPAP-021 | RRID:SAMN19776452 | <!----> | 18M1122 | NA | Female | 21.40 | 13 | Caucasian |  NA | ⋯ |    76 |  3108 | 306 |  1897 |   8954 | NA | NA | T1D  | 10Xv2 | nPod  |\n",
       "| HPAP-022 | HPAP-022 | RRID:SAMN19776453 | <!----> | 18M2405 | NA | Female | 34.70 | 39 | Caucasian | 4.7 | ⋯ |  2788 | 34375 | 437 |  2070 | 315113 | NA | NA | ND   | 10Xv2 | UPenn |\n",
       "| HPAP-023 | HPAP-023 | RRID:SAMN19776454 | <!----> | 18A0312 | NA | Female | 21.35 | 17 | Caucasian | 8.9 | ⋯ |   607 |  3659 | 448 | 25351 |  31682 | NA | NA | T1D  | 10Xv2 | nPod  |\n",
       "| HPAP-024 | HPAP-024 | RRID:SAMN19776455 | <!----> | 18A0813 | NA | Male   | 24.30 | 18 | Caucasian | 5.5 | ⋯ | 29407 |  2183 | 307 |  7435 |  42501 | NA | NA | AAB+ | 10Xv2 | nPod  |\n",
       "\n"
      ],
      "text/plain": [
       "         donor_ID rrid              recovery_opo donor_ID_second unos_ID sex   \n",
       "HPAP-019 HPAP-019 RRID:SAMN19776450              17D0406         NA      Male  \n",
       "HPAP-020 HPAP-020 RRID:SAMN19776451              18J0307         NA      Male  \n",
       "HPAP-021 HPAP-021 RRID:SAMN19776452              18M1122         NA      Female\n",
       "HPAP-022 HPAP-022 RRID:SAMN19776453              18M2405         NA      Female\n",
       "HPAP-023 HPAP-023 RRID:SAMN19776454              18A0312         NA      Female\n",
       "HPAP-024 HPAP-024 RRID:SAMN19776455              18A0813         NA      Male  \n",
       "         bmi   age_years race      hba1c ⋯ beta_cell_count delta_cell_count\n",
       "HPAP-019 29.80 22        Caucasian 5.2   ⋯ 38637            3327           \n",
       "HPAP-020 13.20 14        Caucasian  NA   ⋯   249           14485           \n",
       "HPAP-021 21.40 13        Caucasian  NA   ⋯    76            3108           \n",
       "HPAP-022 34.70 39        Caucasian 4.7   ⋯  2788           34375           \n",
       "HPAP-023 21.35 17        Caucasian 8.9   ⋯   607            3659           \n",
       "HPAP-024 24.30 18        Caucasian 5.5   ⋯ 29407            2183           \n",
       "         epsilon_cell_count pp_cell_count total_events weight height\n",
       "HPAP-019 180                 2114          55007       NA     NA    \n",
       "HPAP-020 939                 6706         202889       NA     NA    \n",
       "HPAP-021 306                 1897           8954       NA     NA    \n",
       "HPAP-022 437                 2070         315113       NA     NA    \n",
       "HPAP-023 448                25351          31682       NA     NA    \n",
       "HPAP-024 307                 7435          42501       NA     NA    \n",
       "         diabetes_status chemistry tissue_source\n",
       "HPAP-019 AAB+            10Xv2     nPod         \n",
       "HPAP-020 T1D             10Xv2     nPod         \n",
       "HPAP-021 T1D             10Xv2     nPod         \n",
       "HPAP-022 ND              10Xv2     UPenn        \n",
       "HPAP-023 T1D             10Xv2     nPod         \n",
       "HPAP-024 AAB+            10Xv2     nPod         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e3bc6a48-dac0-47c9-8df9-d0e9b5bdb1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set cell identities\n",
    "Idents(data) <- data@meta.data$cell_type\n",
    "\n",
    "# Define the samples based on metadata\n",
    "samples <- unique(meta$donor_ID)\n",
    "unique_cell_types <- unique(data$cell_type)\n",
    "\n",
    "# Default assay should be 'RNA'\n",
    "DefaultAssay(data) <- 'RNA'\n",
    "gex.counts <- GetAssayData(data, layer = 'counts')\n",
    "\n",
    "# Create the output directory if it doesn't exist\n",
    "output_dir <- '/nfs/lab/sara/projects/PanKbase/singlecell/analytical_library/output'\n",
    "\n",
    "if (!dir.exists(output_dir)) {\n",
    "  dir.create(output_dir, recursive = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "05d56fef-ba5a-4524-a5d1-7e7488601de4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Unique donors: 65\"\n",
      "[1] \"Unique cell types: 14\"\n"
     ]
    }
   ],
   "source": [
    "print(paste(\"Unique donors:\", length(samples)))\n",
    "print(paste(\"Unique cell types:\", length(unique_cell_types)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8802591-f522-4feb-8560-e613d0247368",
   "metadata": {},
   "source": [
    "## Make Pseudobulk Matrices for each cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bb079bbd-49d5-4b39-9833-4607a2154068",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output directory for cell counts\n",
    "outdir_cell_counts <- file.path(output_dir, \"cell_counts\")\n",
    "dir.create(outdir_cell_counts, showWarnings = FALSE)\n",
    "\n",
    "# Loop over each unique cell type\n",
    "for (cell_type in unique_cell_types) {\n",
    "  # Get barcodes for the current cell type\n",
    "  bcs <- names(Idents(data)[Idents(data) == cell_type])\n",
    "  \n",
    "  # Subset the gene expression counts matrix for the cell type\n",
    "  counts <- gex.counts[, colnames(gex.counts) %in% bcs]\n",
    "  \n",
    "  # Initialize a matrix to store pseudobulk counts for each sample\n",
    "  pseudobulk_counts <- matrix(0, nrow = nrow(gex.counts), ncol = length(samples))\n",
    "  row.names(pseudobulk_counts) <- row.names(gex.counts)\n",
    "  \n",
    "  # Loop over each sample\n",
    "  for (i in seq_along(samples)) {\n",
    "    sample <- samples[i]\n",
    "    \n",
    "    # Get barcodes for the current sample\n",
    "    sample_bcs <- row.names(data@meta.data[data@meta.data$library == sample,])\n",
    "    \n",
    "    # Subset counts for the current sample's barcodes\n",
    "    counts_cut <- counts[, colnames(counts) %in% sample_bcs]\n",
    "    \n",
    "    # Sum the counts across the barcodes for the current sample\n",
    "    if (is.null(ncol(counts_cut)) || ncol(counts_cut) == 0) {\n",
    "      sum_counts <- rep(0, nrow(counts))\n",
    "    } else {\n",
    "      sum_counts <- rowSums(counts_cut)\n",
    "    }\n",
    "      \n",
    "    # Add the summed counts to the pseudobulk matrix\n",
    "    pseudobulk_counts[, i] <- sum_counts\n",
    "  }\n",
    "  \n",
    "  # Convert the matrix to a dataframe and assign sample names as column names\n",
    "  pseudobulk_df <- as.data.frame(pseudobulk_counts)\n",
    "  colnames(pseudobulk_df) <- samples\n",
    "  \n",
    "  # Save the pseudobulk counts\n",
    "  output_file <- file.path(outdir_cell_counts, paste0(cell_type, \"_sample_gex_total_counts.txt\"))\n",
    "  write.table(pseudobulk_df, output_file, sep = '\\t', quote = FALSE, col.names = TRUE, row.names = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a646ad1-7ef2-4189-9738-3bd704b89c5a",
   "metadata": {},
   "source": [
    "## Define helper functions for DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1431977d-b4d9-4c84-afa2-2167dbb4ad0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to read the paths of the count matrices from the dir\n",
    "readCounts <- function(counts_dir) {\n",
    "  # Get all files matching the pattern\n",
    "  files <- list.files(counts_dir, pattern = \"_sample_gex_total_counts.txt$\", full.names = TRUE)\n",
    "  \n",
    "  # Initialize an empty dict to store the cell type and the respective counts matrices' file path\n",
    "  counts_dict <- list()\n",
    "  \n",
    "  # Loop over each file and extract the cell type from the filename\n",
    "  for (file_path in files) {\n",
    "    cell_type <- gsub(\"_sample_gex_total_counts.txt\", \"\", basename(file_path))\n",
    "    \n",
    "    # Read the count matrix and store it in dict\n",
    "    counts_dict[[cell_type]] <- read.csv(paste0(file_path), header = TRUE, row.names=1, sep='\\t', check.names = FALSE)\n",
    "  }\n",
    "  \n",
    "  return(counts_dict)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c7367c78-6610-4c35-abc9-81ea5c9f5cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scale covariates function\n",
    "scaleCovariates <- function(coldata_subset, covariates) {\n",
    "  if (!is.null(covariates)) {\n",
    "    coldata_subset[covariates] <- lapply(coldata_subset[covariates], function(x) {\n",
    "      if (is.numeric(x)) {\n",
    "        return(scale(x))  # Scale numeric covariates\n",
    "      }\n",
    "      return(x)  # Return unchanged for non-numeric covariates\n",
    "    })\n",
    "  }\n",
    "  return(coldata_subset)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "695b104d-f60e-4be4-89fe-e0e6ecb18665",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to filter samples based on the the min number of cells required per sample for a cell type\n",
    "filterSamples  <- function(counts_matrix, coldata, min_cells, cell_type, condition_col, condition_levels) {\n",
    "  # Get cell counts for each sample from the meta data\n",
    "  cell_counts <- table(data@meta.data$library, data@meta.data$cell_type)[, cell_type]\n",
    "  \n",
    "  # Filter out samples that have fewer than min_cells for the given cell type\n",
    "  valid_samples <- names(cell_counts[cell_counts >= min_cells])\n",
    "\n",
    "  # Subset counts matrix and coldata based on valid samples\n",
    "  counts_matrix <- counts_matrix[, colnames(counts_matrix) %in% valid_samples]\n",
    "  coldata <- coldata[rownames(coldata) %in% valid_samples, ]\n",
    "\n",
    "  # Print the number of valid samples for each condition level\n",
    "  for (level in condition_levels) {\n",
    "    samples_in_level <- rownames(coldata[coldata[[condition_col]] == level, ])\n",
    "    num_samples <- length(samples_in_level)\n",
    "    message(paste(\"Number of valid samples in condition level\", level, \"for\", cell_type, \"cells:\", num_samples))\n",
    "  }\n",
    "    \n",
    "  return(list(counts_matrix = counts_matrix, coldata = coldata))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55042807-2c42-4be2-81a5-99837652ce3f",
   "metadata": {},
   "source": [
    "## Perform DESeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c855ce15-67a7-421b-9a0f-b6e008e1452e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the function to perform DESeq2 analysis\n",
    "runDESeq2 <- function(counts_matrix, min_cells, coldata, condition_col, condition_levels, covariates, outdir_deseq, cell_type) {\n",
    "  # Ensure the condition column exists in the coldata\n",
    "  if (!(condition_col %in% colnames(coldata))) {\n",
    "    stop(paste(\"Condition column\", condition_col, \"not found in coldata.\"))\n",
    "  }\n",
    "\n",
    "  # Passing arguments to the filterSamples() function \n",
    "  filtered_data <- filterSamples(counts_matrix, coldata, min_cells, cell_type, condition_col, condition_levels)\n",
    "  counts_matrix <- filtered_data$counts_matrix\n",
    "  coldata <- filtered_data$coldata\n",
    "  \n",
    "  # Ensure there are min of 2 samples in both condition levels\n",
    "  # DESeq2 requires at least two samples per condition level to estimate dispersion \n",
    "  sample_counts <- table(coldata[[condition_col]])\n",
    "  if (any(sample_counts[condition_levels] < 2)) { \n",
    "    message(paste(\"Skipping DESeq2 for\", cell_type, \"- minimum 2 samples required in one or both conditions after sample filtering.\"))\n",
    "    return(NULL)\n",
    "  }\n",
    "\n",
    "  # Subset the coldata for the specified condition levels (e.g., T1D vs ND)\n",
    "  coldata_subset <- coldata[coldata[[condition_col]] %in% condition_levels, ]\n",
    "\n",
    "  # Match the subset of coldata with the counts matrix\n",
    "  counts_subset <- counts_matrix[, rownames(coldata_subset)]\n",
    "    \n",
    "  # Calculate half the sample size for each condition_levels - used as the minimum number of samples to meet gene count thresholds\n",
    "  n_levels <- sapply(condition_levels, function(level) {\n",
    "    sum(coldata_subset[[condition_col]] == level)\n",
    "  })\n",
    "  n_levels_half <- floor(n_levels/2)\n",
    "    \n",
    "  # Filter genes based on the counts criteria - at least half the samples per condition have greater than 5 counts\n",
    "  genes_to_keep <- rownames(counts_subset)[\n",
    "    rowSums(counts_subset[, coldata_subset[[condition_col]] == condition_levels[1]] >= 5) >= n_levels_half[1] &\n",
    "    rowSums(counts_subset[, coldata_subset[[condition_col]] == condition_levels[2]] >= 5) >= n_levels_half[2]\n",
    "  ]\n",
    "\n",
    "  # Subset counts matrix to keep only the genes that passed the filter\n",
    "  counts_subset <- counts_subset[genes_to_keep, ]\n",
    "\n",
    "  # Re-level the factor so the first condition is the reference level\n",
    "  coldata_subset[[condition_col]] <- factor(coldata_subset[[condition_col]], levels = condition_levels)\n",
    "\n",
    "  # Build the design formula with covariates\n",
    "  if (!is.null(covariates) && all(covariates %in% colnames(coldata_subset))) {\n",
    "    design_formula <- as.formula(paste(\"~\", paste(c(covariates, condition_col), collapse = \" + \")))\n",
    "  } else {\n",
    "    design_formula <- as.formula(paste(\"~\", condition_col))\n",
    "  }\n",
    "    \n",
    "  # Create DESeq2 dataset\n",
    "  dds <- DESeqDataSetFromMatrix(countData = counts_subset, colData = coldata_subset, design = design_formula)\n",
    "\n",
    "  # Run DESeq2\n",
    "  dds <- DESeq(dds)\n",
    "\n",
    "  # Extract results for the specified condition comparison\n",
    "  res <- results(dds, contrast = c(condition_col, condition_levels[1], condition_levels[2]))\n",
    "  res <- res[order(res$pvalue), ]\n",
    "\n",
    "  # Save results\n",
    "  condition_comparison <- paste(condition_levels[1], \"vs\", condition_levels[2], sep = \"_\")\n",
    "  output_filename <- paste0(outdir_deseq, \"/\", cell_type, \"_\", condition_comparison, \"_deseq.tsv\")\n",
    "\n",
    "  write.table(res, output_filename, sep = \"\\t\", quote = FALSE, row.names = TRUE, col.names = TRUE)\n",
    "\n",
    "  message(paste(\"Complete! DESeq analysis for\", condition_levels[1], \"vs\", condition_levels[2], \"in\", cell_type, \"cells.\"))\n",
    "\n",
    "  return(dds)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "214392e8-fdda-48ae-b319-805f514731d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_DESeq2_for_all_cell_types <- function(cell_types_list = NULL, min_cells = 10, counts_dir, coldata, condition_col, condition_levels, covariates = NULL, output_dir) {\n",
    "  # Ensure the base output directory exists or create it\n",
    "  outdir_deseq <- file.path(output_dir, \"deseq\")\n",
    "  if (!dir.exists(outdir_deseq)) {\n",
    "    dir.create(outdir_deseq, recursive = TRUE)\n",
    "  }\n",
    "  \n",
    "  # If cell_types_list is not provided, read all count matrices from the directory\n",
    "  if (is.null(cell_types_list)) {\n",
    "    counts_dict <- readCounts(counts_dir)\n",
    "    cell_types_list <- names(counts_dict)  # Use names of the counts_dict as cell types\n",
    "  } else {\n",
    "    counts_dict <- readCounts(counts_dir)[cell_types_list]  # Only read specified cell types\n",
    "  }\n",
    "  \n",
    "  # Loop over each cell type and run DESeq2\n",
    "  for (cell_type in cell_types_list) {\n",
    "    # Retrieve the counts matrix for the current cell type\n",
    "    counts_matrix <- counts_dict[[cell_type]]\n",
    "\n",
    "    # Skip if the counts data for the cell type is missing\n",
    "    if (is.null(counts_matrix)) {\n",
    "      message(paste(\"Skipping DESeq2 for\", cell_type, \"due to missing counts data.\"))\n",
    "      next\n",
    "    }\n",
    "    \n",
    "    # Run DESeq2 for the current cell type and save results\n",
    "    runDESeq2(counts_matrix = counts_matrix, min_cells = min_cells, coldata = coldata, condition_col = condition_col, \n",
    "               condition_levels = condition_levels, covariates = covariates, \n",
    "               outdir_deseq = outdir_deseq, cell_type = cell_type)\n",
    "  }\n",
    "  \n",
    "  message(\"DESeq2 analysis completed for specified cell types.\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606523a6-3dea-4969-b954-1edb18afb811",
   "metadata": {},
   "source": [
    "## Custom DESeq Parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "888ce08a-444f-4a4a-a09a-42863a39c7e8",
   "metadata": {},
   "source": [
    "#### Required Arguments\n",
    "- **coldata**: metadata file that includes sample information, conditions and covariates for DESeq analysis\n",
    "- **condition_col**: The column name in coldata that represents the experimental condition to test (e.g., \"diabetes_status\")\n",
    "- **condition_levels**: A vector specifying the two condition levels to compare in DESeq (e.g., c(\"T1D\", \"ND\"))\n",
    "- **counts_dir**: The directory where the pseudobulk matrices for each cell type will be saved\n",
    "- **output_dir**: The directory where DESeq output will be saved\n",
    "\n",
    "#### Optional Arguments\n",
    "- **cell_types_list**: (default = NULL) list of specific cell types of interest to run DESeq on. If not provided, all cell types in counts_dir are used.\n",
    "- **covariates**: (default = NULL) A list of additional covariates from coldata to include in the DESeq model (e.g., age, sex)\n",
    "- **min_cells**: (default = 10) sets threshold for the minimum number of cells required per sample for a cell type. Samples with fewer cells are removed from analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab812f40-bd31-474a-8250-7649e02669a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of valid samples in condition level T1D for Acinar cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Acinar cells: 27\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "1 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Acinar cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Active_Stellate cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Active_Stellate cells: 26\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Active_Stellate cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Alpha cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Alpha cells: 27\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Alpha cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Alpha+Beta cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Alpha+Beta cells: 25\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Alpha+Beta cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Beta cells: 9\n",
      "\n",
      "Number of valid samples in condition level ND for Beta cells: 27\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Beta cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Cycling_Alpha cells: 3\n",
      "\n",
      "Number of valid samples in condition level ND for Cycling_Alpha cells: 9\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "4 rows did not converge in beta, labelled in mcols(object)$betaConv. Use larger maxit argument with nbinomWaldTest\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Cycling_Alpha cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Delta cells: 8\n",
      "\n",
      "Number of valid samples in condition level ND for Delta cells: 25\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Delta cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Ductal cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Ductal cells: 27\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Ductal cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Endothelial cells: 10\n",
      "\n",
      "Number of valid samples in condition level ND for Endothelial cells: 26\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Endothelial cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Gamma+Epsilon cells: 9\n",
      "\n",
      "Number of valid samples in condition level ND for Gamma+Epsilon cells: 23\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Gamma+Epsilon cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Macrophage cells: 9\n",
      "\n",
      "Number of valid samples in condition level ND for Macrophage cells: 24\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Macrophage cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Mast cells: 8\n",
      "\n",
      "Number of valid samples in condition level ND for Mast cells: 19\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Mast cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for MUC5B+_Ductal cells: 9\n",
      "\n",
      "Number of valid samples in condition level ND for MUC5B+_Ductal cells: 17\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in MUC5B+_Ductal cells.\n",
      "\n",
      "Number of valid samples in condition level T1D for Quiescent_Stellate cells: 9\n",
      "\n",
      "Number of valid samples in condition level ND for Quiescent_Stellate cells: 24\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  the design formula contains one or more numeric variables with integer values,\n",
      "  specifying a model with increasing fold change for higher values.\n",
      "  did you mean for this to be a factor? if so, first convert\n",
      "  this variable to a factor using the factor() function\n",
      "\n",
      "  the design formula contains one or more numeric variables that have mean or\n",
      "  standard deviation larger than 5 (an arbitrary threshold to trigger this message).\n",
      "  Including numeric variables with large mean can induce collinearity with the intercept.\n",
      "  Users should center and scale numeric variables in the design to improve GLM convergence.\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "Complete! DESeq analysis for T1D vs ND in Quiescent_Stellate cells.\n",
      "\n",
      "DESeq2 analysis completed for specified cell types.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "### Required\n",
    "coldata <- meta  # Sample metadata with condition and covariates\n",
    "condition_col <- \"diabetes_status\"\n",
    "condition_levels <- c(\"T1D\", \"ND\")\n",
    "counts_dir <- \"/nfs/lab/sara/projects/PanKbase/singlecell/analytical_library/output/cell_counts\"\n",
    "output_dir <- \"/nfs/lab/sara/projects/PanKbase/singlecell/analytical_library/output\" \n",
    "\n",
    "### Optional\n",
    "cell_types_list <- c(\"Beta\")\n",
    "covariates <- c(\"age_years\", \"sex\", \"bmi\", \"chemistry\", \"tissue_source\")\n",
    "min_cells <- 5\n",
    "\n",
    "# # To run DESeq2 for specific cell types\n",
    "# run_DESeq2_for_all_cell_types(cell_types_list = cell_types_list, \n",
    "#                               coldata = coldata,\n",
    "#                               condition_col = condition_col, \n",
    "#                               condition_levels = condition_levels, \n",
    "#                               counts_dir = counts_dir, \n",
    "#                               output_dir = output_dir,                              \n",
    "#                               min_cells = min_cells,                               \n",
    "#                               covariates = covariates)\n",
    "\n",
    "\n",
    "# To run DESeq2 for all cell types - entire cell_counts dir\n",
    "run_DESeq2_for_all_cell_types(coldata = coldata,\n",
    "                              condition_col = condition_col, \n",
    "                              condition_levels = condition_levels, \n",
    "                              counts_dir = counts_dir, \n",
    "                              output_dir = output_dir,                              \n",
    "                              min_cells = min_cells,                               \n",
    "                              covariates = covariates)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R PanKbase sc DESeq",
   "language": "R",
   "name": "r_pankbase_sc_deseq"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
